src/mono/mono/mini/aot-compiler.c

```diff
@@ -5060,10 +5060,42 @@ MONO_RESTORE_WARNING
 				if (cattr->attrs [j].ctor && mono_is_corlib_image (m_class_get_image (cattr->attrs [j].ctor->klass)) && !strcmp (m_class_get_name (cattr->attrs [j].ctor->klass), "UnmanagedCallersOnlyAttribute"))
 					break;
 			if (j < cattr->num_attrs) {
-				MonoMethod *wrapper = mono_marshal_get_managed_wrapper (method, NULL, 0, error);
+				MonoCustomAttrEntry *e = &cattr->attrs [j];
+				const char *named;
+				int slen;
+				char *export_name = NULL;
+				MonoMethod *wrapper;
+
+				if (!(method->flags & METHOD_ATTRIBUTE_STATIC)) {
+					g_warning ("AOT restriction: Method '%s' must be static since it is decorated with [UnmanagedCallers].",
```

Comment:
Is this permanent limitation?

Reply:
It's part of the [spec of `UnmanagedCallersOnlyAttribute`](https://github.com/dotnet/csharplang/blob/7cf8d52977233ee1995e48c7028723fa433460ce/proposals/csharp-9.0/function-pointers.md#systemruntimeinteropservicesunmanagedcallersonlyattribute)
> It is an error to apply the attribute to anything other than an ordinary static method or ordinary static local function. The C# compiler will mark any non-static or static non-ordinary methods imported from metadata with this attribute as unsupported by the language.

Reply:
Why do we check only for this restriction then and shouldn't this be TLE or something like that?

Reply:
Its just a copy paste from the MonoPInvokeCallback case.